src/codegen/generators/expression_generator.rs

```diff
@@ -522,99 +528,251 @@ impl<'ink, 'b> ExpressionCodeGenerator<'ink, 'b> {
+        let or_value = self.llvm.builder.build_or(and_value, value, "or");
+        self.llvm.builder.build_store(left_pointer, or_value);
 
-                    //Special string handling
-                    if (assigned_output_type.is_string() && output_value_type.is_string())
-                        || (assigned_output_type.is_struct() && output_value_type.is_struct())
-                        || (assigned_output_type.is_array() && output_value_type.is_array())
-                    {
-                        self.generate_string_store(
-                            assigned_output,
-                            assigned_output_type,
-                            expression.get_location(),
-                            output,
-                            output_value_type,
-                            parameter.source_location.clone(),
-                        )?;
-                    } else {
-                        let output_value = builder.build_load(output, "");
-                        builder.build_store(assigned_output, output_value);
-                    }
+        Ok(())
+    }
+
+    fn generate_output_assignment_with_direct_access(
+        &self,
+        left_statement: &AstNode,
+        left_pointer: PointerValue,
+        right_pointer: PointerValue,
+        right_type: &DataType,
+    ) -> Result<(), Diagnostic> {
+        let left_value = self.llvm.builder.build_load(left_pointer, "").into_int_value();
+
+        //Generate an expression for the right size
+        let right = self.llvm.builder.build_load(right_pointer, "");
+        self.generate_assignment_with_direct_access(
+            left_statement,
+            left_value,
+            left_pointer,
+            right_type,
+            right,
+        )?;
+
+        Ok(())
+    }
+
+    fn generate_output_assignment(&self, context: &CallParameterAssignment) -> Result<(), Diagnostic> {
+        let &CallParameterAssignment { assignment: expr, function_name, index, parameter_struct } = context;
+        let builder = &self.llvm.builder;
+
+        // We don't want to generate any code if the right side of an assignment is empty, e.g. `foo(out =>)`
+        if expr.is_empty_statement() {
+            return Ok(());
+        }
+
+        let parameter = self.index.get_declared_parameter(function_name, index).expect("must exist");
+
+        match expr.get_stmt() {
+            AstStatement::ReferenceExpr(_) if expr.has_direct_access() => {
+                // TODO: Can we this be simplified?
+                let rhs_type = {
+                    let pou = self.index.find_pou(function_name).unwrap();
+                    let pou_struct = &pou.find_instance_struct_type(self.index).unwrap().information;
+                    let DataTypeInformation::Struct { members, .. } = pou_struct else { panic!() };
+
+                    self.index.find_effective_type_by_name(&members[index as usize].data_type_name).unwrap()
+                };
```

Comment:
Intended, because I'd argue these have to exist at codegen and otherwise expose a bug in our resolver / validator. Should I introduce a top-level comment / use `expect` to make it more transparent?

Reply:
I'm fine with the unwraps, I think these are just a matter of taste. I would change the `panic` into an `unreachable` if you want to convey that this line is impossible to hit, otherwise make it a diagnostic.